### PR TITLE
Add Value Function and corresponding example script to Diffuser implementation

### PR DESCRIPTION
**Pull request by bglick13**
Builds upon #105 by adding the code necessary for "value guided planning" described in the paper.

Adds an implementation of the Value Function model that adheres to Diffusers api, as well as an example script that shows how to perform value guided planning.

Also included a helper script that converts from the pretrained models provided in the original repo to Diffusers models

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Hey @bglick13, 

Super cool PR :-) @nateraw could you review here? :-)

**Comment by patrickvonplaten** [comment]
BTW @natolambert,

We can also just merge a `unet_rl.py` file in a first step and the before we release we merge the models, layers to not block contributors here - what do you think?

**Comment by natolambert** [comment]
@patrickvonplaten my main question is if the RL-only model is useful to you. It'll allow people to more easily use diffusers to re-train this algorithm (rather than just inference), but it's very niche.

We could consider having it in a script or a community pipeline? 
I've been following the code and working with @bglick13.

**Comment by bglick13** [comment]
Thanks @patrickvonplaten! Like Nathan said the main question is where the RL stuff fits into the HF api. I'm cool with whatever makes the most sense. Just excited to get this code in people's hands. 

I do think it would be cool to facilitate training instead of just inference, but that could either be a part of the core api or a community pipeline as Nathan suggested.

File: .gitignore [hunk 1]
```diff
@@ -163,4 +163,6 @@ tags
 *.lock
 
 # DS_Store (MacOS)
-.DS_Store
\ No newline at end of file
+.DS_Store
```
Comment (natolambert) [inline .gitignore:166]: @bglick13 to get the PR ready, let's remove things like `hub`. 
We can leave `*.mp4` if we add a pipeline, but add a comment saying its for rl so people know.

File: convert_model.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
+
```
Comment (natolambert) [inline convert_model.py:1]: let's try and consolidate all of these files as much as possible. 
Put the convert model into scripts/ with a longer & more descriptive path. (I can do the same for my RL models)

File: examples/diffuser/helpers.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
+import os
```
Comment (natolambert) [inline examples/diffuser/helpers.py:1]: I would guess just put these in the training file to minimize how many files things are across. Also huge points for getting rid of as many functions as you can (I did some of that in colab).

File: examples/diffuser/run_diffuser.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
+import numpy as np
```
Comment (natolambert) [inline examples/diffuser/run_diffuser.py:1]: probably the core of a pipeline or example.

File: src/diffusers/models/unet_rl.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+        ##
+        self.mid_block1 = ResidualTemporalBlock1D(mid_dim, mid_dim // 2, embed_dim=block_out_channels[0])
+        self.mid_down1 = Downsample1D(mid_dim // 2, use_conv=True)
+        ##
+        self.mid_block2 = ResidualTemporalBlock1D(mid_dim // 2, mid_dim // 4, embed_dim=block_out_channels[0])
+        self.mid_down2 = Downsample1D(mid_dim // 4, use_conv=True)
+        ##
+        fc_dim = mid_dim // 4
+        self.final_block = nn.ModuleList(
```
Comment (natolambert) [inline src/diffusers/models/unet_rl.py:78]: generally remove modulelists (harder to debug)

File: src/diffusers/schedulers/scheduling_ddpm.py [hunk 6]
```diff
@@ -283,7 +283,13 @@ def step(
             noise = torch.randn(
                 model_output.size(), dtype=model_output.dtype, layout=model_output.layout, generator=generator
             ).to(model_output.device)
-            variance = (self._get_variance(t, predicted_variance=predicted_variance) ** 0.5) * noise
+            if self.variance_type == "fixed_small_log":
```
Comment (natolambert) [inline src/diffusers/schedulers/scheduling_ddpm.py:286]: can we change the get_variance function to accommodate this? I think the fixed_small_log was added for this?

**Review by natolambert — COMMENTED** [review]
Maybe we can move all of the scripts into a folder with multiple scripts. Not sure if it's a pipeline or what. 

OR

pipelines for training / eval / inference each (also would be nice).

File: examples/diffuser/run_diffuser.py (hunk 4 above)
Comment (bglick13) [reply examples/diffuser/run_diffuser.py:1]: do you want a pipeline to be part of this PR? Or something we add in a small follow up PR?

File: src/diffusers/models/unet_rl.py (hunk 5 above)
Comment (bglick13) [reply src/diffusers/models/unet_rl.py:78]: I'm okay with changing the `final_block` to something cleaner - whatever y'all decide

File: examples/community/pipeline.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+                    grad = torch.autograd.grad([y.sum()], [x])[0]
+
+                    posterior_variance = self.scheduler._get_variance(i)
+                    model_std = torch.exp(0.5 * posterior_variance)
+                    grad = model_std * grad
+                grad[timesteps < 2] = 0
+                x = x.detach()
+                x = x + scale * grad
+                x = self.reset_x0(x, conditions, self.action_dim)
+            # with torch.no_grad():
```
Comment (natolambert) [inline examples/community/pipeline.py:76]:
```suggestion
```

File: examples/community/pipeline.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+            x_in[:, key, act_dim:] = val.clone()
+        return x_in
+
+    def run_diffusion(self, x, conditions, n_guide_steps, scale):
+        batch_size = x.shape[0]
+        y = None
+        for i in tqdm.tqdm(self.scheduler.timesteps):
+            # create batch of timesteps to pass into model
+            timesteps = torch.full((batch_size,), i, device=self.unet.device, dtype=torch.long)
+            # 3. call the sample function
```
Comment (natolambert) [inline examples/community/pipeline.py:62]: Numbering starts at 3, can we remove the comments?

File: examples/community/pipeline.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+                    grad = model_std * grad
+                grad[timesteps < 2] = 0
+                x = x.detach()
+                x = x + scale * grad
+                x = self.reset_x0(x, conditions, self.action_dim)
+            # with torch.no_grad():
+            prev_x = self.unet(x.permute(0, 2, 1), timesteps).sample.permute(0, 2, 1)
+            x = self.scheduler.step(prev_x, i, x, predict_epsilon=False)["prev_sample"]
+
+            # 4. apply conditions to the trajectory
```
Comment (natolambert) [inline examples/community/pipeline.py:80]: fix comment

File: examples/community/pipeline.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+            x = self.reset_x0(x, conditions, self.action_dim)
+            x = self.to_torch(x)
+        # y = network(x, timesteps).sample
```
Comment (natolambert) [inline examples/community/pipeline.py:83]:
```suggestion
```

File: examples/diffuser/run_diffuser.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+action_dim = env.action_space.shape[0]
+num_inference_steps = 100  # number of difusion steps
+
+
+# Two generators for different parts of the diffusion loop to work in colab
+generator_cpu = torch.Generator(device="cpu")
+
+scheduler = DDPMScheduler(num_train_timesteps=100, beta_schedule="squaredcos_cap_v2")
+
+# 3 different pretrained models are available for this task.
```
Comment (natolambert) [inline examples/diffuser/run_diffuser.py:33]: clean up comments

File: examples/diffuser/run_diffuser.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+network = UNet1DModel.from_pretrained("fusing/ddpm-unet-rl-hopper-hor128").to(device=DEVICE)
+# network = TemporalUNet.from_pretrained("fusing/ddpm-unet-rl-hopper-hor256").to(device=DEVICE)
+# network = TemporalUNet.from_pretrained("fusing/ddpm-unet-rl-hopper-hor512").to(device=DEVICE)
+
+
+# network specific constants for inference
+clip_denoised = network.clip_denoised
+predict_epsilon = network.predict_epsilon
+
+# add a batch dimension and repeat for multiple samples
```
Comment (natolambert) [inline examples/diffuser/run_diffuser.py:44]:
```suggestion
```

File: examples/diffuser/run_diffuser_value_guided.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    eta=0.0,
+    t_grad_cutoff=2,
+    device="cpu",
+)
+
+
+def _run():
+    env_name = "hopper-medium-v2"
+    env = gym.make(env_name)
+    # data = env.get_dataset()  # dataset is only used for normalization in this colab
```
Comment (natolambert) [inline examples/diffuser/run_diffuser_value_guided.py:25]:
```suggestion
```

**Review by natolambert — COMMENTED** [review]
this is on the right track. Can we delete the not needed files and clean up some comments? I was going to do more but didn't really know how to clone the remote PR on PR stuff....

File: examples/diffuser/run_diffuser.py [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+import d4rl  # noqa
+import gym
+import tqdm
+import train_diffuser
+from diffusers import DDPMScheduler, UNet1DModel
+
+
+env_name = "hopper-medium-expert-v2"
+env = gym.make(env_name)
+data = env.get_dataset()  # dataset is only used for normalization in this colab
```
[the remark this answers is not in the document]
Comment (bglick13) [reply examples/diffuser/run_diffuser.py:13]: oh actually for value guided I do it in the pipeline. I should convert the pipeline to also work for regular (non value guided diffusion)

**Pull request merged**